### ISSUE TITLE
Deploying on Heroku

### ISSUE THREAD

**Issue by tvarner** [coryhouse/react-slingshot#263]
Node version: v5.0.0

npm version: 3.3.6

Operating system: OSX

Command line used: Terminal

Steps to reproduce: Clone repot. Run `npm run setup`. Deploy react sling-shot on Heroku. Using the four steps from the Heroku Getting Started Tutorial: https://devcenter.heroku.com/articles/getting-started-with-nodejs#deploy-the-app

First off, thanks for the starter kit. It's been a massive help.

I was wondering if anyone had any experience deploying react-slingshot to Heroku. I've built two apps on top of it. Both of which I found had memory leaks after running `$ heroku logs --tail` after trying to deploy. I wasn't sure where the memory leaks were coming from, so I started with just trying to deploy the example react-slingshot app. This also failed for what looks like a different reason. 

I'm relatively new to Heroku. If anyone who has deployed on Heroku using react-slingshot could point me in the right direction, you'd be a hero. Thanks :)

When I log the details of launching react-slingshot, I get the following output:

`$ heroku logs --tail
2016-09-20T21:52:41.722414+00:00 heroku[api]: Enable Logplex by thomas.g.varner@gmail.com
2016-09-20T21:52:41.722414+00:00 heroku[api]: Release v2 created by thomas.g.varner@gmail.com
2016-09-20T21:53:21.115338+00:00 heroku[api]: Scale to web=1 by thomas.g.varner@gmail.com
2016-09-20T21:53:21.116298+00:00 heroku[api]: Deploy 884be3c by thomas.g.varner@gmail.com
2016-09-20T21:53:21.116404+00:00 heroku[api]: Release v3 created by thomas.g.varner@gmail.com
2016-09-20T21:53:21.603620+00:00 heroku[slug-compiler]: Slug compilation started
2016-09-20T21:53:21.603625+00:00 heroku[slug-compiler]: Slug compilation finished
2016-09-20T21:53:22.582516+00:00 heroku[web.1]: Starting process with command `npm start`
2016-09-20T21:53:24.423356+00:00 app[web.1]: > example_react_app@0.1.0 prestart /app
2016-09-20T21:53:24.423342+00:00 app[web.1]: 
2016-09-20T21:53:24.423357+00:00 app[web.1]: > npm-run-all --parallel start-message remove-dist
2016-09-20T21:53:24.423358+00:00 app[web.1]: 
2016-09-20T21:53:24.428706+00:00 app[web.1]: sh: 1: npm-run-all: not found
2016-09-20T21:53:24.438975+00:00 app[web.1]: npm ERR! node v5.11.1
2016-09-20T21:53:24.432801+00:00 app[web.1]: 
2016-09-20T21:53:24.438743+00:00 app[web.1]: npm ERR! Linux 3.13.0-95-generic
2016-09-20T21:53:24.438884+00:00 app[web.1]: npm ERR! argv "/app/.heroku/node/bin/node" "/app/.heroku/node/bin/npm" "start"
2016-09-20T21:53:24.439208+00:00 app[web.1]: npm ERR! code ELIFECYCLE
2016-09-20T21:53:24.439045+00:00 app[web.1]: npm ERR! npm  v3.10.7
2016-09-20T21:53:24.439124+00:00 app[web.1]: npm ERR! file sh
2016-09-20T21:53:24.439293+00:00 app[web.1]: npm ERR! errno ENOENT
2016-09-20T21:53:24.439364+00:00 app[web.1]: npm ERR! syscall spawn
2016-09-20T21:53:24.439433+00:00 app[web.1]: npm ERR! example_react_app@0.1.0 prestart: `npm-run-all --parallel start-message remove-dist`
2016-09-20T21:53:24.439699+00:00 app[web.1]: npm ERR! Make sure you have the latest version of node.js and npm installed.
2016-09-20T21:53:24.439494+00:00 app[web.1]: npm ERR! spawn ENOENT
2016-09-20T21:53:24.439870+00:00 app[web.1]: npm ERR! Tell the author that this fails on your system:
2016-09-20T21:53:24.439572+00:00 app[web.1]: npm ERR! 
2016-09-20T21:53:24.439646+00:00 app[web.1]: npm ERR! Failed at the example_react_app@0.1.0 prestart script 'npm-run-all --parallel start-message remove-dist'.
2016-09-20T21:53:24.439760+00:00 app[web.1]: npm ERR! If you do, this is most likely a problem with the example_react_app package,
2016-09-20T21:53:24.439816+00:00 app[web.1]: npm ERR! not with npm itself.
2016-09-20T21:53:24.439922+00:00 app[web.1]: npm ERR!     npm-run-all --parallel start-message remove-dist
2016-09-20T21:53:24.440030+00:00 app[web.1]: npm ERR! You can get information on how to open an issue for this project with:
2016-09-20T21:53:24.440082+00:00 app[web.1]: npm ERR!     npm bugs example_react_app
2016-09-20T21:53:24.440132+00:00 app[web.1]: npm ERR! Or if that isn't available, you can get their info via:
2016-09-20T21:53:24.443565+00:00 app[web.1]: 
2016-09-20T21:53:24.440182+00:00 app[web.1]: npm ERR!     npm owner ls example_react_app
2016-09-20T21:53:24.440233+00:00 app[web.1]: npm ERR! There is likely additional logging output above.
2016-09-20T21:53:24.443670+00:00 app[web.1]: npm ERR! Please include the following file with any support request:
2016-09-20T21:53:24.443727+00:00 app[web.1]: npm ERR!     /app/npm-debug.log
2016-09-20T21:53:24.485426+00:00 heroku[web.1]: Process exited with status 1
2016-09-20T21:53:24.499145+00:00 heroku[web.1]: State changed from starting to crashed
2016-09-20T21:53:24.500363+00:00 heroku[web.1]: State changed from crashed to starting
2016-09-20T21:53:26.160823+00:00 heroku[web.1]: Starting process with command `npm start`
2016-09-20T21:53:28.528762+00:00 app[web.1]: 
2016-09-20T21:53:28.528778+00:00 app[web.1]: > example_react_app@0.1.0 prestart /app
2016-09-20T21:53:28.528779+00:00 app[web.1]: > npm-run-all --parallel start-message remove-dist
2016-09-20T21:53:28.528780+00:00 app[web.1]: 
2016-09-20T21:53:28.535445+00:00 app[web.1]: sh: 1: npm-run-all: not found
2016-09-20T21:53:28.540004+00:00 app[web.1]: 
2016-09-20T21:53:28.546468+00:00 app[web.1]: npm ERR! Linux 3.13.0-95-generic
2016-09-20T21:53:28.546625+00:00 app[web.1]: npm ERR! argv "/app/.heroku/node/bin/node" "/app/.heroku/node/bin/npm" "start"
2016-09-20T21:53:28.546748+00:00 app[web.1]: npm ERR! node v5.11.1
2016-09-20T21:53:28.546854+00:00 app[web.1]: npm ERR! npm  v3.10.7
2016-09-20T21:53:28.546967+00:00 app[web.1]: npm ERR! file sh
2016-09-20T21:53:28.547075+00:00 app[web.1]: npm ERR! code ELIFECYCLE
2016-09-20T21:53:28.547179+00:00 app[web.1]: npm ERR! errno ENOENT
2016-09-20T21:53:28.547283+00:00 app[web.1]: npm ERR! syscall spawn
2016-09-20T21:53:28.547372+00:00 app[web.1]: npm ERR! example_react_app@0.1.0 prestart: `npm-run-all --parallel start-message remove-dist`
2016-09-20T21:53:28.547536+00:00 app[web.1]: npm ERR! 
2016-09-20T21:53:28.547445+00:00 app[web.1]: npm ERR! spawn ENOENT
2016-09-20T21:53:28.547626+00:00 app[web.1]: npm ERR! Failed at the example_react_app@0.1.0 prestart script 'npm-run-all --parallel start-message remove-dist'.
2016-09-20T21:53:28.547784+00:00 app[web.1]: npm ERR! If you do, this is most likely a problem with the example_react_app package,
2016-09-20T21:53:28.547859+00:00 app[web.1]: npm ERR! not with npm itself.
2016-09-20T21:53:28.547701+00:00 app[web.1]: npm ERR! Make sure you have the latest version of node.js and npm installed.
2016-09-20T21:53:28.547936+00:00 app[web.1]: npm ERR! Tell the author that this fails on your system:
2016-09-20T21:53:28.548008+00:00 app[web.1]: npm ERR!     npm-run-all --parallel start-message remove-dist
2016-09-20T21:53:28.548082+00:00 app[web.1]: npm ERR! You can get information on how to open an issue for this project with:
2016-09-20T21:53:28.548155+00:00 app[web.1]: npm ERR!     npm bugs example_react_app
2016-09-20T21:53:28.548227+00:00 app[web.1]: npm ERR! Or if that isn't available, you can get their info via:
2016-09-20T21:53:28.548300+00:00 app[web.1]: npm ERR!     npm owner ls example_react_app
2016-09-20T21:53:28.548376+00:00 app[web.1]: npm ERR! There is likely additional logging output above.
2016-09-20T21:53:28.552172+00:00 app[web.1]: 
2016-09-20T21:53:28.552332+00:00 app[web.1]: npm ERR! Please include the following file with any support request:
2016-09-20T21:53:28.552404+00:00 app[web.1]: npm ERR!     /app/npm-debug.log
2016-09-20T21:53:28.639797+00:00 heroku[web.1]: State changed from starting to crashed
2016-09-20T21:53:28.624904+00:00 heroku[web.1]: Process exited with status 1`

**Comment by dmurawsky** [comment]
I actually just deployed a couple days ago. The easiest way I found was to put your /dist directory inside a "/dist-server" directory and change the build configuration to build to that /dist, preserving the /dist-server. Then simply add an Express server setup to server the files from /dist. 

Here's the folder structure:
<img width="299" alt="screen shot 2016-09-21 at 11 34 52 am" src="https://cloud.githubusercontent.com/assets/3521396/18717866/743435b6-7fef-11e6-94b0-3ea336931f53.png">

And here's a simple Express server: http://blog.modulus.io/nodejs-and-express-static-content

Then you'll need to change "webpack.config.prod.js" output->path to this:
`output: {
    path: `${__dirname}/dist-server/dist`,
    publicPath: '/',
    filename: '[name].[chunkhash].js'
  },`

And your build scripts in package.json to this:
`    "clean-dist": "npm run remove-dist && mkdir dist-server/dist",
    "remove-dist": "rimraf ./dist-server/dist",
    "prebuild": "npm run clean-dist && npm run lint",
    "build": "babel-node tools/build.js",`

**Comment by tvarner** [comment]
Hi @dmurawsky . Thanks so much for the detailed response. Unfortunately, it wasn't enough to get it working. My package.json and webpack prod config look very similar to the ones provided in react-slingshot. For now, I'm just trying to deploy the react-slingshot example app. And if I can do that, I should hopefully be able to deploy my own app that I've built on top of react-slingshot.

I've implemented your instructions for the example app. If you could put together a screen cap of the process, that'd be amazing. Thanks so much!

My package.json looks like this: 

```{
  "name": "example_react_app",
  "version": "0.1.0",
  "description": "",
  "engines": {
    "npm": ">=3"
  },
  "scripts": {
    "preinstall": "node tools/nodeVersionCheck.js",
    "setup": "node tools/setup/setupMessage.js && npm install && node tools/setup/setup.js",
    "remove-demo": "babel-node tools/removeDemo.js",
    "start-message": "babel-node tools/startMessage.js",
    "prestart": "npm-run-all --parallel start-message remove-dist",
    "start": "npm-run-all --parallel test:watch open:src lint:watch",
    "prod": "NODE_ENV=production node server.js",
    "open:src": "babel-node tools/srcServer.js",
    "open:dist": "babel-node tools/distServer.js",
    "lint": "esw webpack.config.* src tools --color",
    "lint:watch": "npm run lint -- --watch",
    "clean-dist": "npm run remove-dist && mkdir dist-server/dist",
    "remove-dist": "rimraf ./dist-server/dist",
    "prebuild": "npm run clean-dist && npm run lint",
    "build": "babel-node tools/build.js",
    "test": "mocha tools/testSetup.js \"src/**/*.spec.js\" --reporter progress",
    "test:cover": "babel-node node_modules/isparta/bin/isparta cover --root src --report html node_modules/mocha/bin/_mocha -- --require ./tools/testSetup.js \"src/**/*.spec.js\" --reporter progress",
    "test:cover:travis": "babel-node node_modules/isparta/bin/isparta cover --root src --report lcovonly _mocha -- --require ./tools/testSetup.js \"src/**/*.spec.js\" && cat ./coverage/lcov.info | node_modules/coveralls/bin/coveralls.js",
    "test:watch": "npm run test -- --watch",
    "open:cover": "npm run test:cover && open coverage/index.html"
  },
  "author": "",
  "license": "MIT",
  "dependencies": {
    "object-assign": "4.1.0",
    "react": "15.3.1",
    "react-dom": "15.3.1",
    "react-redux": "4.4.5",
    "react-router": "2.7.0",
    "react-router-redux": "4.0.5",
    "redux": "3.5.2",
    "redux-thunk": "2.1.0"
  },
  "devDependencies": {
    "autoprefixer": "6.4.0",
    "babel-cli": "6.14.0",
    "babel-core": "6.14.0",
    "babel-loader": "6.2.5",
    "babel-plugin-react-display-name": "2.0.0",
    "babel-plugin-transform-react-constant-elements": "6.9.1",
    "babel-plugin-transform-react-remove-prop-types": "0.2.9",
    "babel-preset-latest": "6.14.0",
    "babel-preset-react": "6.11.1",
    "babel-preset-react-hmre": "1.1.1",
    "babel-preset-stage-1": "6.13.0",
    "babel-register": "6.14.0",
    "browser-sync": "2.14.0",
    "chai": "3.5.0",
    "chalk": "1.1.3",
    "connect-history-api-fallback": "1.3.0",
    "coveralls": "2.11.12",
    "cross-env": "2.0.1",
    "css-loader": "0.24.0",
    "enzyme": "2.4.1",
    "eslint": "3.4.0",
    "eslint-plugin-import": "1.14.0",
    "eslint-plugin-jsx-a11y": "2.2.0",
    "eslint-plugin-react": "6.2.0",
    "eslint-watch": "2.1.14",
    "extract-text-webpack-plugin": "1.0.1",
    "file-loader": "0.9.0",
    "html-webpack-plugin": "2.22.0",
    "isparta": "4.0.0",
    "mocha": "3.0.2",
    "mockdate": "1.0.4",
    "node-sass": "3.8.0",
    "npm-run-all": "3.0.0",
    "open": "0.0.5",
    "postcss-loader": "0.11.0",
    "prompt": "1.0.0",
    "react-addons-test-utils": "15.3.1",
    "redux-immutable-state-invariant": "1.2.3",
    "replace": "0.3.0",
    "rimraf": "2.5.4",
    "sass-loader": "4.0.0",
    "sinon": "1.17.5",
    "sinon-chai": "2.8.0",
    "style-loader": "0.13.1",
    "url-loader": "0.5.7",
    "webpack": "1.13.2",
    "webpack-dev-middleware": "1.6.1",
    "webpack-hot-middleware": "2.12.2",
    "webpack-md5-hash": "0.0.5"
  },
  "keywords": [],
  "repository": {
    "type": "git",
    "url": ""
  }
}```

and my webpack.config.prod.js looks like this: 

```// For info about this file refer to webpack and webpack-hot-middleware documentation
// For info on how we're generating bundles with hashed filenames for cache busting: https://medium.com/@okonetchnikov/long-term-caching-of-static-assets-with-webpack-1ecb139adb95#.w99i89nsz
import webpack from 'webpack';
import ExtractTextPlugin from 'extract-text-webpack-plugin';
import WebpackMd5Hash from 'webpack-md5-hash';
import HtmlWebpackPlugin from 'html-webpack-plugin';
import autoprefixer from 'autoprefixer';
import path from 'path';

const GLOBALS = {
  'process.env.NODE_ENV': JSON.stringify('production'),
  __DEV__: false
};

export default {
  resolve: {
    extensions: ['', '.js', '.jsx']
  },
  debug: true,
  devtool: 'source-map', // more info:https://webpack.github.io/docs/build-performance.html#sourcemaps and https://webpack.github.io/docs/configuration.html#devtool
  noInfo: true, // set to false to see a list of every file being bundled.
  entry: path.resolve(__dirname, 'src/index'),
  target: 'web', // necessary per https://webpack.github.io/docs/testing.html#compile-and-test
  output: {
    path: path.resolve(__dirname, 'dist-server/dist'),
    publicPath: '/',
    filename: '[name].[chunkhash].js'
  },
  plugins: [
    // Hash the files using MD5 so that their names change when the content changes.
    new WebpackMd5Hash(),

    // Optimize the order that items are bundled. This assures the hash is deterministic.
    new webpack.optimize.OccurenceOrderPlugin(),

    // Tells React to build in prod mode. https://facebook.github.io/react/downloads.html
    new webpack.DefinePlugin(GLOBALS),

    // Generate an external css file with a hash in the filename
    new ExtractTextPlugin('[name].[contenthash].css'),

    // Generate HTML file that contains references to generated bundles. See here for how this works: https://github.com/ampedandwired/html-webpack-plugin#basic-usage
    new HtmlWebpackPlugin({
      template: 'src/index.ejs',
      minify: {
        removeComments: true,
        collapseWhitespace: true,
        removeRedundantAttributes: true,
        useShortDoctype: true,
        removeEmptyAttributes: true,
        removeStyleLinkTypeAttributes: true,
        keepClosingSlash: true,
        minifyJS: true,
        minifyCSS: true,
        minifyURLs: true
      },
      inject: true,
      // Note that you can add custom options here if you need to handle other custom logic in index.html
      // To track JavaScript errors via TrackJS, sign up for a free trial at TrackJS.com and enter your token below.
      trackJSToken: ''
    }),

    // Eliminate duplicate packages when generating bundle
    new webpack.optimize.DedupePlugin(),

    // Minify JS
    new webpack.optimize.UglifyJsPlugin()
  ],
  module: {
    loaders: [
      {test: /\.jsx?$/, exclude: /node_modules/, loader: 'babel'},
      {test: /\.eot(\?v=\d+.\d+.\d+)?$/, loader: 'url?name=[name].[ext]'},
      {test: /\.woff(2)?(\?v=[0-9]\.[0-9]\.[0-9])?$/, loader: "url?limit=10000&mimetype=application/font-woff&name=[name].[ext]"},
      {test: /\.ttf(\?v=\d+.\d+.\d+)?$/, loader: 'url?limit=10000&mimetype=application/octet-stream&name=[name].[ext]'},
      {test: /\.svg(\?v=\d+.\d+.\d+)?$/, loader: 'url?limit=10000&mimetype=image/svg+xml&name=[name].[ext]'},
      {test: /\.(jpe?g|png|gif)$/i, loader: 'file?name=[name].[ext]'},
      {test: /\.ico$/, loader: 'file?name=[name].[ext]'},
      {test: /(\.css|\.scss)$/, loader: ExtractTextPlugin.extract('css?sourceMap!postcss!sass?sourceMap')}
    ]
  },
  web: node server.js,
  postcss: ()=> [autoprefixer]
};```

**Comment by kwelch** [comment]
👍  to @dmurawsky, I have a very similar setup (not used for heroku, I use Azure).

The idea is that you have an express server that knows how to serve static files from the `dist-server/dist`. Additionally, I have modified my `open:dist` script to actually start the express server inside of `dist-server`. 

Key things to note: 
- items inside of `dist-server` should be committed in the repo, with the exception of the `dist` & `node_modules` folder since it is the compiled code
- This is just will only get you to the point of hosting your static files with node, you will still need to find out how to deploy/start a node server within your host/deployment environment.

**Comment by tvarner** [comment]
@kwelch Hi Kyle. Thanks a lot for the tip. This is good info. As soon as I get something working I'll reply to this thread with what I did. Thanks for your time :)

**Comment by garretteklof** [comment]
@kwelch perhaps this isn't the best place to ask this, but I'm looking to deploy my react-slingshot application to Heroku too. I guess I'm a little confused on the folder structure. What is going into `dist-server/dist-server`? As is, the slingshot build script runs 'babel-node' which isn't a production dependency, and without compiling the `dist` folder first, and committing it to repo, I can't seem to see how to compile the assets for an Express server to serve in a platform production environment. Obviously, I am able to do it in the local production environment, because `dist` is built first (via 'babel-node'. Should I refactor the `tools/build.js` to ES5? Then just run it on Heroku with a 'node' command? 

Basically, I have everything working, but don't know how I _should_ build the actual `dist` folder on Heroku, as opposed to hard-compiling it into repo (or using some task runner like Gulp)

I know I am clearly missing something. Thank you in advance!

**Comment by kwelch** [comment]
I have not used Heroku, but I can explain the `dist-server` folder structure and why I am using at my work place. 

Some notes on my approach:
- **This is an example, not a recommendation**
- I use this setup to deploy to Azure
- `dist-server` is committed in the main repo

```
.                                  # root of slingshot
`-src/                         # compiled source code
`-dist-server/            # replaces 'dist' in Slingshot
    `--app/                  # webpack configure to bundle to this directory
    `--package.json   # server package.json - deployed to server
    `-- index.js           # express server configuration
    `-- .gitignore         # mostly used to ensure I don't 'push' node_modules
    `-- web.config     # used in Azure to configure node
```

If you are familiar with Azure, I also have a PowerShell that uses the `.PublishSettings` files to clone the Azure repo and move the `post-build` compiled version of the `dist-server` to be committed and deployed. 

Like I said I am not using Heroku, but I hope you can find some value in this.

**Comment by garretteklof** [comment]
Ah, okay. That's essentially what I ended up doing. I think I was just overthinking it. It's easy, with these un-opinionated systems, to get caught up in best practices, and the "am I really doing this right?" especially when using an (awesome) boilerplate template. It's a process. Thanks so much for sharing your workflow!

**Comment by kwelch** [comment]
Completely, agree I still feel like it is convoluted, but it works and for me at the end of the day that is what matters. Glad it helped!

**Comment by dmurawsky** [comment]
The dist-server/dist-server was a git pull problem.

**Comment by dmurawsky** [comment]
The big problem I ran into was the HTML5 URLs not working: [See this issue for my hacky fix](https://github.com/coryhouse/react-slingshot/issues/273)

If anyone has a better way, please let me know

**Comment by kwelch** [comment]
how are you serving dist-server?

I am using express. 
```
var express = require('express');
var path = require('path');
var compression = require('compression');

var port = process.env.port || 3000;
var app = express();

app.use(compression()); 
app.use(express.static(path.join(__dirname, 'app')));

app.get('*', (_req, res) => {
  res.sendFile(path.join(__dirname, 'app/index.html'));
});

app.listen(port, (err) => {
  if (err) {
    console.log(err);
  } else {
    console.log(`server started port: ${port}`);
  }
});
```

**Comment by dmurawsky** [comment]
```javascript
var express = require('express');
var serveStatic = require('serve-static');
var request = require('request');

app = express();
app.use(serveStatic(__dirname+"/dist"));
var port = process.env.PORT || 5000;

app.route('/*').get(function(req, res, next) {
    res.redirect('/?q='+req.url);
});

app.listen(port);
console.log('server started '+ port);
```

**Comment by kwelch** [comment]
Why are you doing `/*` and redirecting? Feels like it would be a recursive loop for based on your code in #273, since browserHistory is going to trigger a new redirect.

**Comment by kwelch** [comment]
@tvarner Were you able to get this working? Are we able to close this out?

**Comment by kwelch** [comment]
Closing. If there are any other questions, feel free to reach out.

**Comment by rizkysyazuli** [comment]
interesting setup guys. i'm currently struggling with plain nginx and looking for other alternatives.

@kwelch is [that](#issuecomment-252310016) your actual index.js contents? is the package.json inside the `dist-server` folder the same as project's package.json?

**Comment by kwelch** [comment]
That is the `server.js` inside of my `dist-folder`. The package.json is completely different in the folder and only have the relevant dependencies of my express server. 

There was discussion over adding to the documentation a how to deploy section. I have created #408 as a new thread to organize our thoughts and document the different deployment sections.

**Comment by itajenglish** [comment]
I recommend everyone use this build pack for deployment. It's a simple 123 setup works great in production. https://github.com/heroku/heroku-buildpack-static#configuration

**Comment by cassiopagnoncelli** [comment]
@kwelch Works fine here thanks buddy!

**Comment by chenop** [comment]
@kwelch Is there any update on deploying express to Heroku?
Its seems to me very weird to have double package.json and server.js.
what should be inside each one?
Are there any more duplications?

**Comment by kwelch** [comment]
There is not a large amount of duplication. It all depends on how complex of a node server you need. 

The output from react-slingshot are static files so they simply need to be served. If you need complexity that will dictate how much you have to do. 

For my office setup, I have centralized the server configuration between dev and prod into a npm package and then each has necessary differences. 

I hope this helps.

**Comment by guilatrova** [comment]
I just [wrote an article](http://www.latrovacommits.com/en/2017/12/14/how-publish-dist-folder-heroku/) explaining two ways to deploy React Slingshot to Heroku. If you just want code you can check [repository here](https://github.com/guilatrova/deploy-react-slingshot). Hope can help :)

**Comment by kwelch** [comment]
@guilatrova Could you put up a PR adding this to our docs?

**Comment by guilatrova** [comment]
@kwelch Of course, but how you want me to do this? Just add a link to repository, or a link to blog post? If you want me to copy+paste steps, I need to add source code.

**Comment by kwelch** [comment]
I would just add a link to your post. I think that would suite both parties best.

**Comment by chenop** [comment]
Was struggling around this issue - Solved it by doing the following changes in package.json:
```
"scripts: {
...
"start": "node server.js",
"dev": "npm-run-all --parallel open:src lint:watch start-server ",
}
```

My heroku is running the `npm run start` and I'm developing with `npm run dev`.

**Comment by guilatrova** [comment]
@chenop You probably forgot about heroku **Procfile** step. But no problem, you can use script if you want. Only drawback is that you can't use npm run start to work and test as dev.

**Comment by cesc1989** [comment]
> I just [wrote an article](http://www.latrovacommits.com/en/2017/12/14/how-publish-dist-folder-heroku/) explaining two ways to deploy React Slingshot to Heroku. If you just want code you can check [repository here](https://github.com/guilatrova/deploy-react-slingshot). Hope can help :)

This helped me out. Thanks a lot! @guilatrova
